--- esound-0.2.24/configure.in.orig	Fri Mar  1 23:07:05 2002
+++ esound-0.2.24/configure.in	Thu Apr 25 00:39:56 2002
@@ -81,8 +81,7 @@
 
 dnl check for getopt_long, substitute the distributed versions if not
 
-AC_CHECK_FUNC(getopt_long,,[LIBOBJS="$LIBOBJS getopt.o getopt1.o"])
-AC_SUBST(LIBOBJS)
+AC_CHECK_FUNC(getopt_long,,[AC_LIBOBJ([getopt]); AC_LIBOBJ([getopt1])])
 
 dnl Needed for FIONBIO on Solaris at least
 
